### PR TITLE
rgw/sal: LCHead and LCEntry don't need abstraction

### PR DESCRIPTION
**Pull request by cbodley**
(based on https://github.com/ceph/ceph/pull/58448 for now)

`LCHead` and `LCEntry` are just plain old data types. `rgw::sal::Lifecycle` provides all the abstraction we need for zipper

discussed in https://github.com/ceph/ceph/pull/46879#discussion_r1667116923

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
jenkins test api

**Comment by adamemerson** [comment]
It looks like it's failing compilation in dbstore with a difference in SAL interface.

**Comment by adamemerson** [comment]
@mattbenjamin @dang Mind giving this a look? Thank you.

File: src/rgw/driver/dbstore/common/dbstore.cc [hunk 1]
```diff
@@ -2063,7 +2051,7 @@ int DB::set_entry(const std::string& oid, rgw::sal::Lifecycle::LCEntry& entry)
 }
 
 int DB::list_entries(const std::string& oid, const std::string& marker,
-  				 uint32_t max_entries, std::vector<std::unique_ptr<rgw::sal::Lifecycle::LCEntry>>& entries)
+                     uint32_t max_entries, std::vector<rgw::sal::LCEntry>& entries)
```
Comment (mattbenjamin) [inline src/rgw/driver/dbstore/common/dbstore.cc:2054]: std::vector of entries (expanded)?

**Review by mattbenjamin — APPROVED** [review]
I like the simplification.  I assume the vector of LCEntry ops isn't adding overhead?  Seems like you've removed allocations.

File: src/rgw/driver/dbstore/common/dbstore.cc (hunk 1 above)
Comment (cbodley) [reply src/rgw/driver/dbstore/common/dbstore.cc:2054]: right, this just unwraps the individual entries from unique pointers

**Review by dang — APPROVED** [review]

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-09-06_21:44:42-rgw-wip-cbodley-testing-distro-default-smithi/

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
jenkins test make check

**Pull request merged**